code/modules/modular_computers/file_system/programs/antagonist/contract_uplink.dm

```diff
@@ -0,0 +1,133 @@
+			if (traitor_data)
+				hard_drive.traitor_data = traitor_data
+			else
+				error = "Incorrect login details."
+			return 1
+		if("PRG_call_extraction")
+			if (hard_drive.traitor_data.current_contract.status != CONTRACT_STATUS_EXTRACTING)
+				if (hard_drive.traitor_data.current_contract.handle_extraction(user))
+					user.playsound_local(user, 'sound/effects/confirmdropoff.ogg', 75, 1)
+					hard_drive.traitor_data.current_contract.status = CONTRACT_STATUS_EXTRACTING
+				else
+					user.playsound_local(user, 'sound/machines/uplinkerror.ogg', 50)
+			else 
+				user.playsound_local(user, 'sound/machines/uplinkerror.ogg', 50)
+			
+			return 1
+		if("PRG_contract_abort")
+			var/contract_id = hard_drive.traitor_data.current_contract.id
+
+			hard_drive.traitor_data.current_contract = null
+			hard_drive.traitor_data.assigned_contracts[contract_id].status = CONTRACT_STATUS_ABORTED
+
+			return 1
+		if("PRG_redeem_TC")
+			if (hard_drive.traitor_data.contract_TC_to_redeem)
+				var/obj/item/stack/telecrystal/crystals = new /obj/item/stack/telecrystal(get_turf(user), 
+															hard_drive.traitor_data.contract_TC_to_redeem)
+				if(ishuman(user))
+					var/mob/living/carbon/human/H = user
+					if(H.put_in_hands(crystals))
+						to_chat(H, "Your payment materializes into your hands!")
+					else
+						to_chat(user, "Your payment materializes onto the floor.")
+
+				hard_drive.traitor_data.contract_TC_payed_out += hard_drive.traitor_data.contract_TC_to_redeem
+				hard_drive.traitor_data.contract_TC_to_redeem = 0
+				return 1
+			else
+				user.playsound_local(user, 'sound/machines/uplinkerror.ogg', 50)
+			return 1
+
+
+/datum/computer_file/program/contract_uplink/ui_data(mob/user)
+	var/list/data = list()
+	var/obj/item/computer_hardware/hard_drive/small/syndicate/hard_drive = computer.all_components[MC_HDD]
+
+	if (hard_drive && hard_drive.traitor_data != null)
+		var/datum/antagonist/traitor/traitor_data = hard_drive.traitor_data
+
+		error = ""
+		data = get_header_data()
+
+		if (traitor_data.current_contract)
+			data["ongoing_contract"] = TRUE
+			if (traitor_data.current_contract.status == CONTRACT_STATUS_EXTRACTING)
+				data["extraction_enroute"] = TRUE
+
+		data["logged_in"] = TRUE
+		data["station_name"] = GLOB.station_name
+		data["redeemable_tc"] = traitor_data.contract_TC_to_redeem
+
+		for (var/datum/syndicate_contract/contract in traitor_data.assigned_contracts)
+			data["contracts"] += list(list(
+				"target" = contract.contract.target,
+				"payout" = contract.contract.payout,
+				"payout_bonus" = contract.contract.payout_bonus,
+				"dropoff" = contract.contract.dropoff,
+				"id" = contract.id,
+				"status" = contract.status
+			))
+
+		var/direction
+		if (traitor_data.current_contract)
+			var/turf/curr = get_turf(user)
+			var/turf/dropoff_turf 
+			data["current_location"] = "[get_area_name(curr, TRUE)]"
+			
+			for (var/content in traitor_data.current_contract.contract.dropoff.contents)
```

Comment:
Areas have z property that's lowest z among their turfs so you don't really need to iterate here from what i see.

Reply:
This loop is for finding the first instance of a turf inside the area, since .contents gives turfs as well as obj. Is there a better way to get that from what you're saying?

Reply:
Just `area_instance.z` is enough.